go/vt/vttablet/tabletmanager/filesystemmanager.go

```diff
@@ -0,0 +1,131 @@
+package tabletmanager
+
+import (
+	"context"
+	"os"
+	"path"
+	"strconv"
+	"sync"
+	"time"
+)
+
+type FileSystemManager interface {
+	// IsDiskStalled returns true if the disk is stalled or rejecting writes.
+	IsDiskStalled() bool
+}
+
+func newFileSystemManager(ctx context.Context) FileSystemManager {
+	if stalledDiskWriteDir == "" {
+		return newNoopFilesystemManager()
+	}
+
+	return newPollingFileSystemManager(ctx, attemptFileWrite, stalledDiskWriteInterval, stalledDiskWriteTimeout)
+}
+
+type writeFunction func() error
+
+func attemptFileWrite() error {
+	file, err := os.Create(path.Join(stalledDiskWriteDir, ".stalled_disk_check"))
+	if err != nil {
+		return err
+	}
+	_, err = file.WriteString(strconv.FormatInt(time.Now().UnixNano(), 10))
+	if err != nil {
+		return err
+	}
+	err = file.Sync()
+	if err != nil {
+		return err
+	}
+	return file.Close()
+}
+
+type pollingFileSystemManager struct {
+	stalledMutex         sync.RWMutex
+	stalled              bool
+	writeInProgressMutex sync.RWMutex
+	writeInProgress      bool
+	writeFunc            writeFunction
+	pollingInterval      time.Duration
+	writeTimeout         time.Duration
+}
+
+var _ FileSystemManager = &pollingFileSystemManager{}
+
+func newPollingFileSystemManager(ctx context.Context, writeFunc writeFunction, pollingInterval, writeTimeout time.Duration) *pollingFileSystemManager {
+	fs := &pollingFileSystemManager{
+		stalledMutex:         sync.RWMutex{},
+		stalled:              false,
+		writeInProgressMutex: sync.RWMutex{},
+		writeInProgress:      false,
+		writeFunc:            writeFunc,
+		pollingInterval:      pollingInterval,
+		writeTimeout:         writeTimeout,
+	}
+	go fs.poll(ctx)
+	return fs
+}
+
+func (fs *pollingFileSystemManager) poll(ctx context.Context) {
```

Comment:
Interesting -- I don't see why this check wouldn't work if you set `--stalled-disk-write-dir` to the same directory as `--log_dir` in the event that directory could not be written to. Is your suggestion to run the stalled disk check against the `--log_dir` in addition to whatever data directory the user provides with `--stalled-disk-write-dir`? Or even to be able to provide multiple directories as arguments and poll each separately?

Reply:
yeah, I was wondering if other threads that might be writing to disk directly could cause other parts of vitess to be I/O blocked, specially if we are logging something. setting both `--log_dir` and `--stalled-disk-write-dir` to directories in the same filesystem and freezing it would be a good additional test, in case you are only logging to `STDERR` right now. if you are already writing to the same filesystem and this is working, great!

Reply:
I think the main goal for VTOrc detections is to prevent availability/durability problems, so if the log dir being unwritable doesn't impact query serving I'm on the fence if that would be a reason to failover

I'm curious what @GuptaManan100 thinks about this detail?

Reply:
I agree with @timvaillancourt. Not being able to log probably shouldn't trigger an automatic failover. If that is a use-case we want to address, I think it would be better to add a metric around not being able to log, and let the users see the metric on vttablet page and if they deem it worthwhile, they can run the ERS manually.

Reply:
I agree with that, my question is specifically towards if by enabling `--logdir` and having other threads (including the ones that can impact availability) be blocked because they are trying to log something and the flags makes it go to disk instead (or in addition) to `STDERR`. if the disk is stalled, they might be blocked as well. I am not worried if we stop logging because the disk is stalled but `VTOrc` still does the right thing.

Reply:
yeah, I don't mean to block this at all, I think it is a good change to have (we had to have our automation to handle this particular case). But just wanted to mention it _might_ still stall and prevent some failover under some conditions, and it would be good to audit those (not necessarily as part of this PR)